File: src/parser/JeebParser.ts
import { TransactionParser } from './TransactionParser';
import { ParseResult, TransactionType } from '../types';

export class JeebParser implements TransactionParser {
  private readonly validSenders = ['Jeeb', 'PremierBank', 'Premier', 'JeebPay'];

  canParse(sender: string, body: string): boolean {
    const s = sender.trim().toLowerCase();
    return this.validSenders.some((v) => s.includes(v.toLowerCase())) || body.toLowerCase().includes('jeeb');
  }

  parse(sender: string, body: string): ParseResult | null {
    try {
      let type: TransactionType = 'EXPENSE';
      let amount = 0;
      let counterparty: string | undefined;
      let balance: number | undefined;
      let reference: string | undefined;

      const lower = body.toLowerCase();
      if (lower.includes('received') || lower.includes('waxaad heshay') || lower.includes('credited')) {
        type = 'INCOME';
      } else {
        type = 'EXPENSE';
      }

      const amountMatch = body.match(/\$\s*([0-9.,]+)/) || body.match(/USD\s*([0-9.,]+)/i) || body.match(/([0-9.,]+)\s*USD/i);
      if (amountMatch) {
        amount = parseFloat(amountMatch[1].replace(/,/g, ''));
      }
      if (isNaN(amount) || amount <= 0) return null;

      const toFromMatch = body.match(/(?:from|to|ka|u)\s+([^.\n,]+)/i);
      if (toFromMatch) {
        counterparty = toFromMatch[1].trim();
      }

      const balMatch = body.match(/balance[:\s]*\$?\s*([0-9.,]+)/i) || body.match(/haraaga[:\s]*\$?\s*([0-9.,]+)/i);
      if (balMatch) {
        const b = parseFloat(balMatch[1].replace(/,/g, ''));
        if (!isNaN(b)) balance = b;
      }

      const refMatch = body.match(/(?:TxId|Ref|Reference)[:\s]*([A-Za-z0-9]+)/i);
      if (refMatch) {
        reference = refMatch[1].trim();
      }

      return {
        amount,
        type,
        sender: type === 'INCOME' ? counterparty : undefined,
        receiver: type !== 'INCOME' ? counterparty : undefined,
        balance,
        reference,
        provider: 'Jeeb',
        category: type === 'INCOME' ? 'Dakhli' : 'Other',
        confidenceScore: 0.9,
        timestamp: Date.now(),
      };
    } catch {
      return null;
    }
  }
}

File: app/src/main/java/com/example/smsreaderpro/data/parser/JeebParser.kt
package com.example.smsreaderpro.data.parser

import com.example.smsreaderpro.data.model.ParseResult
import com.example.smsreaderpro.data.model.TransactionType
import java.util.Locale

class JeebParser : TransactionParser {
    private val validSenders = listOf("Jeeb", "PremierBank", "Premier", "JeebPay")

    override fun canParse(sender: String, body: String): Boolean {
        val s = sender.trim().lowercase(Locale.ROOT)
        return validSenders.any { s.contains(it.lowercase(Locale.ROOT)) } || body.contains("jeeb", ignoreCase = true)
    }

    override fun parse(sender: String, body: String): ParseResult? {
        return try {
            val lower = body.lowercase(Locale.ROOT)
            val type = if (lower.contains("received") || lower.contains("waxaad heshay") || lower.contains("credited")) {
                TransactionType.INCOME
            } else {
                TransactionType.EXPENSE
            }

            val amountRegex = Regex("""\$\s*([0-9.,]+)""")
            val amountMatch = amountRegex.find(body) ?: Regex("""USD\s*([0-9.,]+)""", RegexOption.IGNORE_CASE).find(body)
            val amount = amountMatch?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull() ?: 0.0

            if (amount <= 0.0) return null

            val toFromMatch = Regex("""(?:from|to|ka|u)\s+([^.\n,]+)""", RegexOption.IGNORE_CASE).find(body)
            val counterparty = toFromMatch?.groupValues?.get(1)?.trim()

            val balMatch = Regex("""(?:balance|haraaga)[:\s]*\$?\s*([0-9.,]+)""", RegexOption.IGNORE_CASE).find(body)
            val balance = balMatch?.groupValues?.get(1)?.replace(",", "")?.toDoubleOrNull()

            val refMatch = Regex("""(?:TxId|Ref|Reference)[:\s]*([A-Za-z0-9]+)""", RegexOption.IGNORE_CASE).find(body)
            val reference = refMatch?.groupValues?.get(1)?.trim()

            ParseResult(
                amount = amount,
                type = type,
                sender = if (type == TransactionType.INCOME) counterparty else null,
                receiver = if (type != TransactionType.INCOME) counterparty else null,
                balance = balance,
                reference = reference,
                provider = "Jeeb",
                category = if (type == TransactionType.INCOME) "Dakhli" else "Other",
                confidenceScore = 0.9f,
                timestamp = System.currentTimeMillis()
            )
        } catch (_: Exception) {
            null
        }
    }
}
